<a href="https://colab.research.google.com/github/FranciaAlday/Proyectos/blob/main/PLN/Proyectos/Proyecto_NLP_esp.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Francia Alday

**SemEval 2024 Track B Unsupervised Español**

#Español

In [ ]:
import re
import pandas as pd
import numpy as np

In [ ]:
!pip install spacy
!python -m spacy download es_core_news_sm


2023-11-28 05:12:56.256804: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-11-28 05:12:56.256898: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-11-28 05:12:56.256994: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-11-28 05:12:57.652249: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.9/12.9 MB 50.3 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_sm')


In [ ]:
esp = pd.read_csv('esp_dev.csv')
esp.head(10)

PairID                                               Text
0  ESP-dev-0000  Notable es la carta de Robert a Sarah.\r\nStra...
1  ESP-dev-0001  Este tipo de tratamiento para grandes contribu...
2  ESP-dev-0002  El filme es denso y con lentitud nos sumerge e...
3  ESP-dev-0003  Un " catéter de diálisis " es un catéter usado...
4  ESP-dev-0004  Aunque tengamos las mejores intenciones, puede...
5  ESP-dev-0005  SHACL (Shapes Constraint Language) es una espe...
6  ESP-dev-0006  ¿Dónde está Natchitoches, Luisiana?\r\n¿Dónde ...
7  ESP-dev-0007  Actualmente forma parte de la " Ruta Moche ".\...
8  ESP-dev-0008  No es un robot con personalidad propia, pero s...
9  ESP-dev-0009  Mansour Bahrami y Henri Leconte ganaron en la ...

In [ ]:
esp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140 entries, 0 to 139
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   PairID  140 non-null    object
 1   Text    140 non-null    object
dtypes: object(2)
memory usage: 2.3+ KB


trackA = pd.read_csv('dev_esp.csv')
trackA.info()

In [ ]:
print(esp)

           PairID                                               Text
0    ESP-dev-0000  Notable es la carta de Robert a Sarah.\r\nStra...
1    ESP-dev-0001  Este tipo de tratamiento para grandes contribu...
2    ESP-dev-0002  El filme es denso y con lentitud nos sumerge e...
3    ESP-dev-0003  Un " catéter de diálisis " es un catéter usado...
4    ESP-dev-0004  Aunque tengamos las mejores intenciones, puede...
..            ...                                                ...
135  ESP-dev-0135  Mejor cámara pero con cambios menores.\r\nEl ú...
136  ESP-dev-0136  Los vecindarios notables incluyen: McElderry P...
137  ESP-dev-0137  El estudio comenzará con el análisis de todos ...
138  ESP-dev-0138  Un niño nadando en el océano lanza una moneda ...
139  ESP-dev-0139  Un grupo de personas bailan en un concierto.\r...

[140 rows x 2 columns]


In [ ]:
# Dividir la columna 'Text' en dos nuevas columnas
esp[['Frase1', 'Frase2']] = esp['Text'].str.split('\r\n', 1, expand=True)

esp.head()

<ipython-input-66-97f60842c92f>:2: FutureWarning: In a future version of pandas all arguments of StringMethods.split except for the argument 'pat' will be keyword-only.
  esp[['Frase1', 'Frase2']] = esp['Text'].str.split('\r\n', 1, expand=True)


PairID                                               Text  \
0  ESP-dev-0000  Notable es la carta de Robert a Sarah.\r\nStra...   
1  ESP-dev-0001  Este tipo de tratamiento para grandes contribu...   
2  ESP-dev-0002  El filme es denso y con lentitud nos sumerge e...   
3  ESP-dev-0003  Un " catéter de diálisis " es un catéter usado...   
4  ESP-dev-0004  Aunque tengamos las mejores intenciones, puede...   

                                              Frase1  \
0             Notable es la carta de Robert a Sarah.   
1  Este tipo de tratamiento para grandes contribu...   
2  El filme es denso y con lentitud nos sumerge e...   
3  Un " catéter de diálisis " es un catéter usado...   
4  Aunque tengamos las mejores intenciones, puede...   

                                              Frase2  
0  Strathairn asistió a Williams College, William...  
1  Bill Clinton fue el primer presidente negro de...  
2  El talento de Johnny Depp queda maquillado y d...  
3  Si un paciente requiere terapia de diálisis de...  
4  ", le pregunto a Krista Fransman, una amiga fi...

In [ ]:
# Seleccionar las columnas relevantes para el nuevo conjunto de datos
frases_esp = esp[['Frase1', 'Frase2']]
# Guardar el nuevo conjunto de datos en un archivo CSV
#frases_esp.to_csv('nuevo_dataset.csv', index=False)

In [ ]:
frases_esp.head()

Frase1  \
0             Notable es la carta de Robert a Sarah.   
1  Este tipo de tratamiento para grandes contribu...   
2  El filme es denso y con lentitud nos sumerge e...   
3  Un " catéter de diálisis " es un catéter usado...   
4  Aunque tengamos las mejores intenciones, puede...   

                                              Frase2  
0  Strathairn asistió a Williams College, William...  
1  Bill Clinton fue el primer presidente negro de...  
2  El talento de Johnny Depp queda maquillado y d...  
3  Si un paciente requiere terapia de diálisis de...  
4  ", le pregunto a Krista Fransman, una amiga fi...

In [ ]:
# Convertir a minúsculas las columnas 'Frase1' y 'Frase2'
frases_esp['Frase1'] = frases_esp['Frase1'].str.lower()
frases_esp['Frase2'] = frases_esp['Frase2'].str.lower()

<ipython-input-69-d0e921d55d5b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  frases_esp['Frase1'] = frases_esp['Frase1'].str.lower()
<ipython-input-69-d0e921d55d5b>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  frases_esp['Frase2'] = frases_esp['Frase2'].str.lower()


In [ ]:
frases_espp = frases_esp.copy()

# Tokenizar las columnas 'Frase1' y 'Frase2' conservando los signos de puntuación
#frases_espp['Frase1'] = frases_espp['Frase1'].apply(nltk.wordpunct_tokenize)
#frases_espp['Frase2'] = frases_espp['Frase2'].apply(nltk.wordpunct_tokenize)

# TfidfVectorizer

## Con signos de puntuación

TfidfVectorizer no toma en cuenta signos de puntuación para hacer las matrices

In [ ]:
import nltk
from nltk.tokenize import word_tokenize
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

### Así como va

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# Crear el objeto vectorizador TF-IDF. Usaremos unigramas y bigramas.
tfidf = TfidfVectorizer(ngram_range=(1, 2,), token_pattern=r'\b\w+\b|[\.,;\'"¡!\?]')

# Ajustar el vectorizador TF-IDF en todas las oraciones y transformar cada conjunto por separado
matriz_f1 = tfidf.fit_transform(frases_espp['Frase1'])
matriz_f2 = tfidf.transform(frases_espp['Frase2'])

# Calcular la similitud de coseno para cada par de oraciones correspondiente
cosine_similarities = cosine_similarity(matriz_f1, matriz_f2)

# Crear un dataframe con las similitudes coseno
sim_esp_p0c = pd.DataFrame(cosine_similarities.diagonal(), columns=['Cosine Similarity'])

# Adjuntar las similitudes al dataframe original para una mejor visualización
sim_esp_p0c = pd.concat([frases_espp[['Frase1','Frase2']], sim_esp_p0c], axis=1)

In [ ]:
sim_esp_p0c.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Cosine Similarity  
0   strathairn asistió a williams college, william...           0.053648  
1   bill clinton fue el primer presidente negro de...           0.021805  
2   el talento de johnny depp queda maquillado y d...           0.071073  
3   si un paciente requiere terapia de diálisis de...           0.420510  
4   ", le pregunto a krista fransman, una amiga fi...           0.048299  
5   celebra a lo largo del año diversas festividades.           0.004644  
6            ¿dónde está el puente sobre el río kwai?           0.454614  
7   es el balneario más visitado de la ciudad de t...           0.091933  
8   puedes decirlos en alto o escribirlos, lo que ...           0.017073  
9   jeremy bates y anders järryd ganaron contra ma...           0.911895  
10  los materiales de ingeniería más resistentes a...           0.269678  
11  fue el nacimiento del trip-hop y del sonido br...           0.147619  
12  hoy, la asignación de recursos a hospitales pú...           0.076454  
13  allí llegó después de estudiar filosofía en la...           0.026171  
14           así es como, mmm..., me quedo abrochado.           0.412148  
15  algunos pilotos salieron disparados y consigui...           0.071525  
16  fue nominada para un sobey art award (nominada...           0.642825  
17  dos equipos de jugadores de fútbol no comienza...           0.456911  
18  localidad ubicada al oeste de la llamada pampa...           0.481076  
19  ¿cómo gana un padre el respeto de un niño que ...           0.047773

### Con stemming

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import nltk
from nltk.stem import SnowballStemmer
from nltk.tokenize import RegexpTokenizer

# Descargar el recurso necesario para NLTK
nltk.download('punkt')

# Configurar el stemmer para español
stemmer = SnowballStemmer('spanish')

# Tokenización con una expresión regular que incluye signos de puntuación
tokenizer = RegexpTokenizer(r'\w+|[^\w\s]')
frases_espp['Frase1_stemmed_psc'] = frases_espp['Frase1'].apply(lambda x: ' '.join([stemmer.stem(word) for word in tokenizer.tokenize(x)]))
frases_espp['Frase2_stemmed_psc'] = frases_espp['Frase2'].apply(lambda x: ' '.join([stemmer.stem(word) for word in tokenizer.tokenize(x)]))

# Crear el objeto vectorizador TF-IDF. Usaremos unigramas y bigramas.
tfidf = TfidfVectorizer(ngram_range=(1, 2,), token_pattern=r'\b\w+\b|[\.,;\'"¡!\?]')

# Ajustar el vectorizador TF-IDF en todas las oraciones y transformar cada conjunto por separado
matriz_f1 = tfidf.fit_transform(frases_espp['Frase1_stemmed_psc'])
matriz_f2 = tfidf.transform(frases_espp['Frase2_stemmed_psc'])

# Calcular la similitud de coseno para cada par de oraciones correspondiente
cosine_similarities = cosine_similarity(matriz_f1, matriz_f2)

# Crear un dataframe con las similitudes coseno
sim_esp_psc = pd.DataFrame(cosine_similarities.diagonal(), columns=['Cosine Similarity'])

# Adjuntar las similitudes al dataframe original para una mejor visualización
sim_esp_psc = pd.concat([frases_espp[['Frase1','Frase2']], sim_esp_psc], axis=1)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [ ]:
sim_esp_psc.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Cosine Similarity  
0   strathairn asistió a williams college, william...           0.050050  
1   bill clinton fue el primer presidente negro de...           0.022142  
2   el talento de johnny depp queda maquillado y d...           0.061538  
3   si un paciente requiere terapia de diálisis de...           0.421824  
4   ", le pregunto a krista fransman, una amiga fi...           0.049024  
5   celebra a lo largo del año diversas festividades.           0.004085  
6            ¿dónde está el puente sobre el río kwai?           0.427447  
7   es el balneario más visitado de la ciudad de t...           0.094153  
8   puedes decirlos en alto o escribirlos, lo que ...           0.016074  
9   jeremy bates y anders järryd ganaron contra ma...           0.911267  
10  los materiales de ingeniería más resistentes a...           0.276872  
11  fue el nacimiento del trip-hop y del sonido br...           0.136508  
12  hoy, la asignación de recursos a hospitales pú...           0.068075  
13  allí llegó después de estudiar filosofía en la...           0.024448  
14           así es como, mmm..., me quedo abrochado.           0.395466  
15  algunos pilotos salieron disparados y consigui...           0.059242  
16  fue nominada para un sobey art award (nominada...           0.642308  
17  dos equipos de jugadores de fútbol no comienza...           0.460643  
18  localidad ubicada al oeste de la llamada pampa...           0.428914  
19  ¿cómo gana un padre el respeto de un niño que ...           0.044411

### Con Lematización

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import spacy

# Cargar el modelo de spaCy para español
nlp = spacy.load("es_core_news_sm")

# Aplicar lematización a la columna incluyendo signos de puntuación
frases_espp['Frase1_lemma_plc'] = frases_espp['Frase1'].apply(lambda x: ' '.join([token.lemma_ for token in nlp(x)]))
frases_espp['Frase2_lemma_plc'] = frases_espp['Frase2'].apply(lambda x: ' '.join([token.lemma_ for token in nlp(x)]))

# Crear el objeto vectorizador TF-IDF. Usaremos unigramas y bigramas.
tfidf = TfidfVectorizer(ngram_range=(1, 2,), token_pattern=r'\b\w+\b|[\.,;\'"¡!\?]')

# Ajustar el vectorizador TF-IDF en todas las oraciones y transformar cada conjunto por separado
matriz_f1 = tfidf.fit_transform(frases_espp['Frase1_lemma_plc'])
matriz_f2 = tfidf.transform(frases_espp['Frase2_lemma_plc'])

# Calcular la similitud de coseno para cada par de oraciones correspondiente
cosine_similarities = cosine_similarity(matriz_f1, matriz_f2)

# Crear un dataframe con las similitudes coseno
sim_esp_plc = pd.DataFrame(cosine_similarities.diagonal(), columns=['Cosine Similarity'])

# Adjuntar las similitudes al dataframe original para una mejor visualización
sim_esp_plc = pd.concat([frases_espp[['Frase1','Frase2']], sim_esp_plc], axis=1)

In [ ]:
sim_esp_plc.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Cosine Similarity  
0   strathairn asistió a williams college, william...           0.048462  
1   bill clinton fue el primer presidente negro de...           0.033782  
2   el talento de johnny depp queda maquillado y d...           0.069458  
3   si un paciente requiere terapia de diálisis de...           0.412069  
4   ", le pregunto a krista fransman, una amiga fi...           0.045898  
5   celebra a lo largo del año diversas festividades.           0.025521  
6            ¿dónde está el puente sobre el río kwai?           0.479489  
7   es el balneario más visitado de la ciudad de t...           0.084183  
8   puedes decirlos en alto o escribirlos, lo que ...           0.075670  
9   jeremy bates y anders järryd ganaron contra ma...           0.910630  
10  los materiales de ingeniería más resistentes a...           0.306585  
11  fue el nacimiento del trip-hop y del sonido br...           0.124813  
12  hoy, la asignación de recursos a hospitales pú...           0.102512  
13  allí llegó después de estudiar filosofía en la...           0.027162  
14           así es como, mmm..., me quedo abrochado.           0.385282  
15  algunos pilotos salieron disparados y consigui...           0.076539  
16  fue nominada para un sobey art award (nominada...           0.673456  
17  dos equipos de jugadores de fútbol no comienza...           0.421950  
18  localidad ubicada al oeste de la llamada pampa...           0.375004  
19  ¿cómo gana un padre el respeto de un niño que ...           0.053770

## Sin signos de puntuación

### Así como va

#### Similitud Coseno

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# Crear el objeto vectorizador TF-IDF. Usaremos unigramas y bigramas.
tfidf = TfidfVectorizer(ngram_range=(1, 2,))

# Ajustar el vectorizador TF-IDF en todas las oraciones y transformar cada conjunto por separado
matriz_f1_s = tfidf.fit_transform(frases_espp['Frase1'])
matriz_f2_s = tfidf.transform(frases_espp['Frase2'])

# Calcular la similitud de coseno para cada par de oraciones correspondiente
cosine_similarities = cosine_similarity(matriz_f1_s, matriz_f2_s)

# Crear un dataframe con las similitudes coseno
sim_esp_s0c = pd.DataFrame(cosine_similarities.diagonal(), columns=['Cosine Similarity'])

# Adjuntar las similitudes al dataframe
sim_esp_s0c = pd.concat([frases_espp[['Frase1','Frase2']], sim_esp_s0c], axis=1)

In [ ]:
sim_esp_s0c.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Cosine Similarity  
0   strathairn asistió a williams college, william...           0.043362  
1   bill clinton fue el primer presidente negro de...           0.019865  
2   el talento de johnny depp queda maquillado y d...           0.061816  
3   si un paciente requiere terapia de diálisis de...           0.506568  
4   ", le pregunto a krista fransman, una amiga fi...           0.038929  
5   celebra a lo largo del año diversas festividades.           0.000000  
6            ¿dónde está el puente sobre el río kwai?           0.472509  
7   es el balneario más visitado de la ciudad de t...           0.102914  
8   puedes decirlos en alto o escribirlos, lo que ...           0.000000  
9   jeremy bates y anders järryd ganaron contra ma...           0.903336  
10  los materiales de ingeniería más resistentes a...           0.241127  
11  fue el nacimiento del trip-hop y del sonido br...           0.150036  
12  hoy, la asignación de recursos a hospitales pú...           0.057175  
13  allí llegó después de estudiar filosofía en la...           0.025380  
14           así es como, mmm..., me quedo abrochado.           0.000000  
15  algunos pilotos salieron disparados y consigui...           0.098177  
16  fue nominada para un sobey art award (nominada...           0.679365  
17  dos equipos de jugadores de fútbol no comienza...           0.484361  
18  localidad ubicada al oeste de la llamada pampa...           0.523757  
19  ¿cómo gana un padre el respeto de un niño que ...           0.016337

#### Dice Score

In [ ]:
import re

def preprocess_text(text):
    # Eliminar puntuación y convertir a minúsculas
    text = re.sub(r'[^\w\s]', '', text.lower())
    return text

def dice_score(s1, s2):
    s1 = preprocess_text(s1)
    s2 = preprocess_text(s2)

    s1_split = set(re.findall(r"\w+|[^\w\s]", s1, re.UNICODE))
    s2_split = set(re.findall(r"\w+|[^\w\s]", s2, re.UNICODE))

    dice_coef = len(s1_split.intersection(s2_split)) / (len(s1_split) + len(s2_split))
    return round(dice_coef, 2)

# Aplicar la función a cada par de frases en tu DataFrame
dice_scores = [dice_score(row['Frase1'], row['Frase2']) for index, row in frases_espp.iterrows()]

# Crear un dataframe con los coeficientes de Dice
sim_esp_s0d = pd.DataFrame({'Dice Coefficient': dice_scores})

# Adjuntar los coeficientes de Dice al dataframe original
sim_esp_s0d = pd.concat([frases_espp[['Frase1', 'Frase2']], sim_esp_s0d], axis=1)

In [ ]:
sim_esp_s0d.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Dice Coefficient  
0   strathairn asistió a williams college, william...              0.11  
1   bill clinton fue el primer presidente negro de...              0.04  
2   el talento de johnny depp queda maquillado y d...              0.11  
3   si un paciente requiere terapia de diálisis de...              0.17  
4   ", le pregunto a krista fransman, una amiga fi...              0.06  
5   celebra a lo largo del año diversas festividades.              0.00  
6            ¿dónde está el puente sobre el río kwai?              0.18  
7   es el balneario más visitado de la ciudad de t...              0.12  
8   puedes decirlos en alto o escribirlos, lo que ...              0.00  
9   jeremy bates y anders järryd ganaron contra ma...              0.50  
10  los materiales de ingeniería más resistentes a...              0.14  
11  fue el nacimiento del trip-hop y del sonido br...              0.09  
12  hoy, la asignación de recursos a hospitales pú...              0.07  
13  allí llegó después de estudiar filosofía en la...              0.04  
14           así es como, mmm..., me quedo abrochado.              0.00  
15  algunos pilotos salieron disparados y consigui...              0.09  
16  fue nominada para un sobey art award (nominada...              0.31  
17  dos equipos de jugadores de fútbol no comienza...              0.22  
18  localidad ubicada al oeste de la llamada pampa...              0.23  
19  ¿cómo gana un padre el respeto de un niño que ...              0.04

### Con stemming

#### Similitud coseno

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import nltk
from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer
nltk.download('punkt')

# Configurar el stemmer para español
stemmer = SnowballStemmer('spanish')

# Aplicar stemming a las columnas
frases_espp['Frase1_stemmed_ssc'] = frases_espp['Frase1'].apply(lambda x: ' '.join([stemmer.stem(word) for word in word_tokenize(x)]))
frases_espp['Frase2_stemmed_ssc'] = frases_espp['Frase2'].apply(lambda x: ' '.join([stemmer.stem(word) for word in word_tokenize(x)]))

# Crear el objeto vectorizador TF-IDF. Usaremos unigramas y bigramas.
tfidf = TfidfVectorizer(ngram_range=(1, 2,))

# Ajustar el vectorizador TF-IDF en todas las oraciones y transformar cada conjunto por separado
matriz_f1 = tfidf.fit_transform(frases_espp['Frase1_stemmed_ssc'])
matriz_f2 = tfidf.transform(frases_espp['Frase2_stemmed_ssc'])

# Calcular la similitud de coseno para cada par de oraciones correspondiente
cosine_similarities = cosine_similarity(matriz_f1, matriz_f2)

# Crear un dataframe con las similitudes coseno
sim_esp_ssc = pd.DataFrame(cosine_similarities.diagonal(), columns=['Cosine Similarity'])

# Adjuntar las similitudes al dataframe original para una mejor visualización
sim_esp_ssc = pd.concat([frases_espp[['Frase1','Frase2']], sim_esp_ssc], axis=1)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [ ]:
sim_esp_ssc.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Cosine Similarity  
0   strathairn asistió a williams college, william...           0.036765  
1   bill clinton fue el primer presidente negro de...           0.020217  
2   el talento de johnny depp queda maquillado y d...           0.053212  
3   si un paciente requiere terapia de diálisis de...           0.508260  
4   ", le pregunto a krista fransman, una amiga fi...           0.039554  
5   celebra a lo largo del año diversas festividades.           0.000000  
6            ¿dónde está el puente sobre el río kwai?           0.438036  
7   es el balneario más visitado de la ciudad de t...           0.106551  
8   puedes decirlos en alto o escribirlos, lo que ...           0.000000  
9   jeremy bates y anders järryd ganaron contra ma...           0.902209  
10  los materiales de ingeniería más resistentes a...           0.250366  
11  fue el nacimiento del trip-hop y del sonido br...           0.138180  
12  hoy, la asignación de recursos a hospitales pú...           0.050907  
13  allí llegó después de estudiar filosofía en la...           0.023494  
14           así es como, mmm..., me quedo abrochado.           0.000000  
15  algunos pilotos salieron disparados y consigui...           0.075281  
16  fue nominada para un sobey art award (nominada...           0.678866  
17  dos equipos de jugadores de fútbol no comienza...           0.488813  
18  localidad ubicada al oeste de la llamada pampa...           0.466909  
19  ¿cómo gana un padre el respeto de un niño que ...           0.014849

#### Dice Score

In [ ]:
import re
import nltk
from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer
#nltk.download('punkt')

# Configurar el stemmer para español
stemmer = SnowballStemmer('spanish')

# Aplicar stemming a las columnas
frases_espp['Frase1_stemmed_ssd'] = frases_espp['Frase1'].apply(lambda x: ' '.join([stemmer.stem(word) for word in word_tokenize(x)]))
frases_espp['Frase2_stemmed_ssd'] = frases_espp['Frase2'].apply(lambda x: ' '.join([stemmer.stem(word) for word in word_tokenize(x)]))

def preprocess_text(text):
    # Eliminar puntuación y convertir a minúsculas
    text = re.sub(r'[^\w\s]', '', text.lower())
    return text

def dice_score(s1, s2):
    s1 = preprocess_text(s1)
    s2 = preprocess_text(s2)

    s1_split = set(re.findall(r"\w+|[^\w\s]", s1, re.UNICODE))
    s2_split = set(re.findall(r"\w+|[^\w\s]", s2, re.UNICODE))

    dice_coef = len(s1_split.intersection(s2_split)) / (len(s1_split) + len(s2_split))
    return round(dice_coef, 2)

# Aplicar la función a cada par de frases en tu DataFrame
dice_scores = [dice_score(row['Frase1_stemmed_ssd'], row['Frase2_stemmed_ssd']) for index, row in frases_espp.iterrows()]

# Crear un dataframe con los coeficientes de Dice
sim_esp_ssd = pd.DataFrame({'Dice Coefficient': dice_scores})

# Adjuntar los coeficientes de Dice al dataframe original
sim_esp_ssd = pd.concat([frases_espp[['Frase1', 'Frase2']], sim_esp_ssd], axis=1)

In [ ]:
sim_esp_ssd.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Dice Coefficient  
0   strathairn asistió a williams college, william...              0.11  
1   bill clinton fue el primer presidente negro de...              0.04  
2   el talento de johnny depp queda maquillado y d...              0.11  
3   si un paciente requiere terapia de diálisis de...              0.17  
4   ", le pregunto a krista fransman, una amiga fi...              0.06  
5   celebra a lo largo del año diversas festividades.              0.00  
6            ¿dónde está el puente sobre el río kwai?              0.18  
7   es el balneario más visitado de la ciudad de t...              0.12  
8   puedes decirlos en alto o escribirlos, lo que ...              0.00  
9   jeremy bates y anders järryd ganaron contra ma...              0.50  
10  los materiales de ingeniería más resistentes a...              0.14  
11  fue el nacimiento del trip-hop y del sonido br...              0.09  
12  hoy, la asignación de recursos a hospitales pú...              0.07  
13  allí llegó después de estudiar filosofía en la...              0.04  
14           así es como, mmm..., me quedo abrochado.              0.00  
15  algunos pilotos salieron disparados y consigui...              0.09  
16  fue nominada para un sobey art award (nominada...              0.31  
17  dos equipos de jugadores de fútbol no comienza...              0.22  
18  localidad ubicada al oeste de la llamada pampa...              0.23  
19  ¿cómo gana un padre el respeto de un niño que ...              0.04

#### Distancia Euclidiana

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import euclidean_distances
import nltk
from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer

# Configurar el stemmer para español
stemmer = SnowballStemmer('spanish')

# Aplicar stemming a las columnas
frases_espp['Frase1_stemmed_ssc'] = frases_espp['Frase1'].apply(lambda x: ' '.join([stemmer.stem(word) for word in word_tokenize(x)]))
frases_espp['Frase2_stemmed_ssc'] = frases_espp['Frase2'].apply(lambda x: ' '.join([stemmer.stem(word) for word in word_tokenize(x)]))

# Crear el objeto vectorizador TF-IDF. Usaremos unigramas y bigramas.
tfidf = TfidfVectorizer(ngram_range=(1, 2,))

# Ajustar el vectorizador TF-IDF en todas las oraciones y transformar cada conjunto por separado
matriz_f1 = tfidf.fit_transform(frases_espp['Frase1_stemmed_ssc'])
matriz_f2 = tfidf.transform(frases_espp['Frase2_stemmed_ssc'])

# Calcular la distancia euclidiana para cada par de oraciones correspondiente
euclidean_distances = euclidean_distances(matriz_f1, matriz_f2)

# Crear un DataFrame con las distancias euclidianas
dist_esp_sse = pd.DataFrame(euclidean_distances.diagonal(), columns=['Euclidean Distance'])

# Ajustar el parámetro alpha
alpha = 0.001

# Calcular la similitud utilizando la función exponencial inversa
dist_esp_sse['Similitud'] = np.exp(-alpha * dist_esp_sse['Euclidean Distance'])

# Normalizar la similitud a una escala de 0 a 1
dist_esp_sse['Similitud'] = ( (dist_esp_sse['Similitud'] - dist_esp_sse['Similitud'].min()) / (dist_esp_sse['Similitud'].max() - dist_esp_sse['Similitud'].min()) )

# Adjuntar las distancias al DataFrame original para una mejor visualización
dist_esp_sse = pd.concat([frases_espp[['Frase1', 'Frase2']], dist_esp_sse], axis=1)

In [ ]:
dist_esp_sse.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Euclidean Distance  \
0   strathairn asistió a williams college, william...            1.387973   
1   bill clinton fue el primer presidente negro de...            1.399845   
2   el talento de johnny depp queda maquillado y d...            1.376073   
3   si un paciente requiere terapia de diálisis de...            0.991706   
4   ", le pregunto a krista fransman, una amiga fi...            1.385962   
5   celebra a lo largo del año diversas festividades.            1.414214   
6            ¿dónde está el puente sobre el río kwai?            1.060154   
7   es el balneario más visitado de la ciudad de t...            1.336749   
8   puedes decirlos en alto o escribirlos, lo que ...            1.414214   
9   jeremy bates y anders järryd ganaron contra ma...            0.442246   
10  los materiales de ingeniería más resistentes a...            1.224446   
11  fue el nacimiento del trip-hop y del sonido br...            1.312874   
12  hoy, la asignación de recursos a hospitales pú...            1.377747   
13  allí llegó después de estudiar filosofía en la...            1.397502   
14           así es como, mmm..., me quedo abrochado.            1.414214   
15  algunos pilotos salieron disparados y consigui...            1.359940   
16  fue nominada para un sobey art award (nominada...            0.801416   
17  dos equipos de jugadores de fútbol no comienza...            1.011125   
18  localidad ubicada al oeste de la llamada pampa...            1.032561   
19  ¿cómo gana un padre el respeto de un niño que ...            1.403675   

    Similitud  
0    0.018542  
1    0.010153  
2    0.026951  
3    0.298610  
4    0.019963  
5    0.000000  
6    0.250225  
7    0.054739  
8    0.000000  
9    0.687133  
10   0.134104  
11   0.071611  
12   0.025768  
13   0.011808  
14   0.000000  
15   0.038351  
16   0.433140  
17   0.284883  
18   0.269730  
19   0.007447

#### Distancia Jaccard

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import pairwise_distances
import nltk
from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer

# Configurar el stemmer para español
stemmer = SnowballStemmer('spanish')

# Aplicar stemming a las columnas
frases_espp['Frase1_stemmed_ssc'] = frases_espp['Frase1'].apply(lambda x: ' '.join([stemmer.stem(word) for word in word_tokenize(x)]))
frases_espp['Frase2_stemmed_ssc'] = frases_espp['Frase2'].apply(lambda x: ' '.join([stemmer.stem(word) for word in word_tokenize(x)]))

# Crear el objeto vectorizador TF-IDF. Usaremos unigramas y bigramas.
tfidf = TfidfVectorizer(ngram_range=(1, 2,))

# Ajustar el vectorizador TF-IDF en todas las oraciones y transformar cada conjunto por separado
matriz_f1 = tfidf.fit_transform(frases_espp['Frase1_stemmed_ssc'])
matriz_f2 = tfidf.transform(frases_espp['Frase2_stemmed_ssc'])

# Convertir las matrices dispersas a matrices densas
matriz_f1_dense = matriz_f1.toarray()
matriz_f2_dense = matriz_f2.toarray()

# Calcular la distancia de jaccard para cada par de oraciones
jaccard_distances = pairwise_distances(matriz_f1_dense, matriz_f2_dense, metric='jaccard')

# Crear un DataFrame con las distancias
dist_esp_ssj = pd.DataFrame(jaccard_distances.diagonal(), columns=['Jaccard Distance'])

# Crear un DataFrame con las similitudes invertidas de Jaccard
dist_esp_ssj['Jaccard Similarity'] = pd.DataFrame(1 - jaccard_distances.diagonal(), columns=['Jaccard Similarity'])

# Adjuntar las distancias al DataFrame original para una mejor visualización
dist_esp_ssj = pd.concat([frases_espp[['Frase1', 'Frase2']], dist_esp_ssj], axis=1)

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/pairwise.py:2025: DataConversionWarning: Data was converted to boolean for metric jaccard
  warnings.warn(msg, DataConversionWarning)


In [ ]:
dist_esp_ssj.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Jaccard Distance  \
0   strathairn asistió a williams college, william...          0.888889   
1   bill clinton fue el primer presidente negro de...          0.967742   
2   el talento de johnny depp queda maquillado y d...          0.931818   
3   si un paciente requiere terapia de diálisis de...          0.757576   
4   ", le pregunto a krista fransman, una amiga fi...          0.947368   
5   celebra a lo largo del año diversas festividades.          1.000000   
6            ¿dónde está el puente sobre el río kwai?          0.666667   
7   es el balneario más visitado de la ciudad de t...          0.842105   
8   puedes decirlos en alto o escribirlos, lo que ...          1.000000   
9   jeremy bates y anders järryd ganaron contra ma...          0.160000   
10  los materiales de ingeniería más resistentes a...          0.888889   
11  fue el nacimiento del trip-hop y del sonido br...          0.942857   
12  hoy, la asignación de recursos a hospitales pú...          0.959184   
13  allí llegó después de estudiar filosofía en la...          0.958333   
14           así es como, mmm..., me quedo abrochado.          1.000000   
15  algunos pilotos salieron disparados y consigui...          0.900000   
16  fue nominada para un sobey art award (nominada...          0.483871   
17  dos equipos de jugadores de fútbol no comienza...          0.666667   
18  localidad ubicada al oeste de la llamada pampa...          0.687500   
19  ¿cómo gana un padre el respeto de un niño que ...          0.972222   

    Jaccard Similarity  
0             0.111111  
1             0.032258  
2             0.068182  
3             0.242424  
4             0.052632  
5             0.000000  
6             0.333333  
7             0.157895  
8             0.000000  
9             0.840000  
10            0.111111  
11            0.057143  
12            0.040816  
13            0.041667  
14            0.000000  
15            0.100000  
16            0.516129  
17            0.333333  
18            0.312500  
19            0.027778

### Con Lematización

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import spacy

# Cargar el modelo de spaCy para español
nlp = spacy.load("es_core_news_sm")

# Aplicar lematización a la columna
frases_espp['Frase1_lemma_slc'] = frases_espp['Frase1'].apply(lambda x: ' '.join([token.lemma_ for token in nlp(x)]))
frases_espp['Frase2_lemma_slc'] = frases_espp['Frase2'].apply(lambda x: ' '.join([token.lemma_ for token in nlp(x)]))

# Crear el objeto vectorizador TF-IDF. Usaremos unigramas y bigramas.
tfidf = TfidfVectorizer(ngram_range=(1, 2,))

# Ajustar el vectorizador TF-IDF en todas las oraciones y transformar cada conjunto por separado
matriz_f1_s = tfidf.fit_transform(frases_espp['Frase1_lemma_slc'])
matriz_f2_s = tfidf.transform(frases_espp['Frase2_lemma_slc'])

# Calcular la similitud de coseno para cada par de oraciones correspondiente
cosine_similarities_s = cosine_similarity(matriz_f1_s, matriz_f2_s)

# Crear un dataframe con las similitudes coseno
sim_esp_slc = pd.DataFrame(cosine_similarities_s.diagonal(), columns=['Cosine Similarity'])

# Adjuntar las similitudes al dataframe
sim_esp_slc = pd.concat([frases_espp[['Frase1','Frase2']], sim_esp_slc], axis=1)

In [ ]:
sim_esp_slc.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Cosine Similarity  
0   strathairn asistió a williams college, william...           0.032246  
1   bill clinton fue el primer presidente negro de...           0.033037  
2   el talento de johnny depp queda maquillado y d...           0.061645  
3   si un paciente requiere terapia de diálisis de...           0.492079  
4   ", le pregunto a krista fransman, una amiga fi...           0.042747  
5   celebra a lo largo del año diversas festividades.           0.022339  
6            ¿dónde está el puente sobre el río kwai?           0.501805  
7   es el balneario más visitado de la ciudad de t...           0.092832  
8   puedes decirlos en alto o escribirlos, lo que ...           0.075391  
9   jeremy bates y anders järryd ganaron contra ma...           0.901054  
10  los materiales de ingeniería más resistentes a...           0.282198  
11  fue el nacimiento del trip-hop y del sonido br...           0.125345  
12  hoy, la asignación de recursos a hospitales pú...           0.059574  
13  allí llegó después de estudiar filosofía en la...           0.026285  
14           así es como, mmm..., me quedo abrochado.           0.000000  
15  algunos pilotos salieron disparados y consigui...           0.100650  
16  fue nominada para un sobey art award (nominada...           0.704934  
17  dos equipos de jugadores de fútbol no comienza...           0.450003  
18  localidad ubicada al oeste de la llamada pampa...           0.409732  
19  ¿cómo gana un padre el respeto de un niño que ...           0.021755

## La mejor?

Creo que la mejor ha sido sim_esp_ssc = sin puntuación, stemming, cosine similarity

Porque hasta ahora es el que rankea a las frases con index=3 lo más perecidas con 0.508260, así como las frases con index=9. Además de dejar en 0.0 las que no se parecen

In [ ]:
sim_esp_ssc.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Cosine Similarity  
0   strathairn asistió a williams college, william...           0.036765  
1   bill clinton fue el primer presidente negro de...           0.020217  
2   el talento de johnny depp queda maquillado y d...           0.053212  
3   si un paciente requiere terapia de diálisis de...           0.508260  
4   ", le pregunto a krista fransman, una amiga fi...           0.039554  
5   celebra a lo largo del año diversas festividades.           0.000000  
6            ¿dónde está el puente sobre el río kwai?           0.438036  
7   es el balneario más visitado de la ciudad de t...           0.106551  
8   puedes decirlos en alto o escribirlos, lo que ...           0.000000  
9   jeremy bates y anders järryd ganaron contra ma...           0.902209  
10  los materiales de ingeniería más resistentes a...           0.250366  
11  fue el nacimiento del trip-hop y del sonido br...           0.138180  
12  hoy, la asignación de recursos a hospitales pú...           0.050907  
13  allí llegó después de estudiar filosofía en la...           0.023494  
14           así es como, mmm..., me quedo abrochado.           0.000000  
15  algunos pilotos salieron disparados y consigui...           0.075281  
16  fue nominada para un sobey art award (nominada...           0.678866  
17  dos equipos de jugadores de fútbol no comienza...           0.488813  
18  localidad ubicada al oeste de la llamada pampa...           0.466909  
19  ¿cómo gana un padre el respeto de un niño que ...           0.014849

In [ ]:
# Crear un nuevo DataFrame con las columnas "PairID" y "Pred_Score"
File_esp_ssc = pd.concat([esp['PairID'], sim_esp_ssc['Cosine Similarity']], axis=1)

# Cambiar el nombre de la columna "Cosine Similarity" a "Pred_Score"
File_esp_ssc = File_esp_ssc.rename(columns={'Cosine Similarity': 'Pred_Score'})

In [ ]:
File_esp_ssc.head(15)

PairID  Pred_Score
0   ESP-dev-0000    0.036765
1   ESP-dev-0001    0.020217
2   ESP-dev-0002    0.053212
3   ESP-dev-0003    0.508260
4   ESP-dev-0004    0.039554
5   ESP-dev-0005    0.000000
6   ESP-dev-0006    0.438036
7   ESP-dev-0007    0.106551
8   ESP-dev-0008    0.000000
9   ESP-dev-0009    0.902209
10  ESP-dev-0010    0.250366
11  ESP-dev-0011    0.138180
12  ESP-dev-0012    0.050907
13  ESP-dev-0013    0.023494
14  ESP-dev-0014    0.000000

In [ ]:
#File_esp_ssc[['PairID', 'Pred_Score']].round({'Pred_Score': 2}).to_csv('pred_esp_ssc.csv', index=False) #, header=False)

In [ ]:
# Guardar el DataFrame en un archivo CSV
#File_esp_ssc[['PairID','Pred_Score']].to_csv('pred_esp_ssc.csv', index=False)

In [ ]:
#File_esp_ssc[['PairID', 'Pred_Score']].to_csv('pred_esp_ssc.csv', index=False, header=False)

In [ ]:
File_esp_ssc[['PairID', 'Pred_Score']].round({'Pred_Score': 2}).astype({'Pred_Score': str}).applymap(lambda x: f'{float(x):.2f}' if '.' in x else x).to_csv('pred_esp_ssc.csv', index=False)

In [ ]:
File_esp_ssc[['PairID', 'Pred_Score']].round({'Pred_Score': 2}).astype({'Pred_Score': str}).applymap(lambda x: f'{float(x):.2f}' if '.' in x else x).to_csv('pred_esp_ssc.csv', index=False, sep=',')

In [ ]:
pred = pd.read_csv('pred_esp_ssc.csv')
print(pred)

           PairID  Pred_Score
0    ESP-dev-0000        0.04
1    ESP-dev-0001        0.02
2    ESP-dev-0002        0.05
3    ESP-dev-0003        0.51
4    ESP-dev-0004        0.04
..            ...         ...
135  ESP-dev-0135        0.12
136  ESP-dev-0136        0.00
137  ESP-dev-0137        0.03
138  ESP-dev-0138        0.17
139  ESP-dev-0139        0.02

[140 rows x 2 columns]


### Generate submission file

Submission file has two columns: '**PairID**' and '**Pred_Score**'

df_str_rel[['PairID', 'Pred_Score']].to_csv('pred_eng.csv', index=False)

## Conclusiones:

1. Es mejor quitar la puntuación porque si 2 frases tienen algún carácter de puntuación parecido, la similitud puede subir aunque las frases no tengan nada en común.

Con puntuación:

14 |	si ahora puede demostrarlo ... |	así es como, mmm..., me quedo abrochado.	| 0.41214796040152235

Sin puntuación:

14	| si ahora puede demostrarlo ...	| así es como, mmm..., me quedo abrochado. |	0.0




#SpaCy

In [ ]:
import spacy
!python -m spacy download es_dep_news_trf
!pip install spacy-transformers

# Cargar el modelo
#nlps = spacy.load("es_dep_news_trf")

2023-11-28 05:13:28.495803: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-11-28 05:13:28.495884: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-11-28 05:13:28.495934: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-11-28 05:13:31.189836: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 410.2/410.2 MB 1.4 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_dep_news_trf')


In [ ]:
nlps = spacy.load('es_dep_news_trf')

ValueError: ignored

In [ ]:
# Función para calcular la similitud entre frases
def spacy_similarity(s1, s2):
    # Procesar las frases con el modelo de SpaCy
    doc1 = nlps(s1)
    doc2 = nlps(s2)

    # Calcular la similitud entre los documentos
    similarity = doc1.similarity(doc2)

    # Escalar la similitud al rango [0, 1]
    scaled_similarity = (similarity + 1) / 2

    return round(scaled_similarity, 2)

# Calcular la similitud para cada par de frases
similarity_scores_spacy = [spacy_similarity(frases_espp['Frase1'].iloc[i], frases_espp['Frase2'].iloc[i]) for i in range(len(frases_espp))]

# Crear un DataFrame con las puntuaciones de similitud
spacy_sim = pd.DataFrame({'SpaCy Similarity': similarity_scores_spacy})

# Adjuntar las puntuaciones de similitud al DataFrame original
spacy_sim = pd.concat([frases_espp[['Frase1', 'Frase2']], spacy_sim], axis=1)

In [ ]:
spacy_sim.head(20)

# Word2Vec

## BERT

In [ ]:
!pip install sentence-transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 1.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 21.8 MB/s eta 0:00:00
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125923 sha256=36ec208198d9f92c9274d338d4f69fcacb0670bab14405b32bce51b0c28cd92b
  Stored in directory: /root/.cache/pip/wheels/62/f2/10/1e606fd5f02395388f74e7462910fe851042f97238cbbd902f
Successfully built sentence-transformers


In [ ]:
from sentence_transformers import SentenceTransformer, util
import pandas as pd

# Cargar el modelo BERT preentrenado en español
model_name = 'paraphrase-xlm-r-multilingual-v1'
model = SentenceTransformer(model_name)

# Calcular la similitud entre frases
def bert_similarity(s1, s2):
    # Obtener las representaciones vectoriales de las frases
    embeddings = model.encode([s1, s2], convert_to_tensor=True)

    # Calcular la similitud coseno entre las representaciones vectoriales de las frases
    similarity = util.pytorch_cos_sim(embeddings[0], embeddings[1]).item()

    # Ajustar la escala al rango [0, 1]
    scaled_similarity = (similarity + 1) / 2

    return round(scaled_similarity, 2)

# Calcular la similitud para cada par de frases
similarity_scores = [bert_similarity(frases_espp['Frase1'].iloc[i], frases_espp['Frase2'].iloc[i]) for i in range(len(frases_espp))]

# Crear un DataFrame con las puntuaciones de similitud
bert_00c = pd.DataFrame({'BERT Similarity': similarity_scores})

# Adjuntar las puntuaciones de similitud al DataFrame original
bert_00c = pd.concat([frases_espp[['Frase1', 'Frase2']], bert_00c], axis=1)

.gitattributes:   0%|          | 0.00/345 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/3.74k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/718 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.11G [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/150 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.10M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/550 [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

In [ ]:
bert_00c.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  BERT Similarity  
0   strathairn asistió a williams college, william...             0.62  
1   bill clinton fue el primer presidente negro de...             0.54  
2   el talento de johnny depp queda maquillado y d...             0.62  
3   si un paciente requiere terapia de diálisis de...             0.84  
4   ", le pregunto a krista fransman, una amiga fi...             0.51  
5   celebra a lo largo del año diversas festividades.             0.59  
6            ¿dónde está el puente sobre el río kwai?             0.70  
7   es el balneario más visitado de la ciudad de t...             0.70  
8   puedes decirlos en alto o escribirlos, lo que ...             0.59  
9   jeremy bates y anders järryd ganaron contra ma...             0.96  
10  los materiales de ingeniería más resistentes a...             0.86  
11  fue el nacimiento del trip-hop y del sonido br...             0.58  
12  hoy, la asignación de recursos a hospitales pú...             0.59  
13  allí llegó después de estudiar filosofía en la...             0.54  
14           así es como, mmm..., me quedo abrochado.             0.68  
15  algunos pilotos salieron disparados y consigui...             0.61  
16  fue nominada para un sobey art award (nominada...             0.97  
17  dos equipos de jugadores de fútbol no comienza...             0.79  
18  localidad ubicada al oeste de la llamada pampa...             0.70  
19  ¿cómo gana un padre el respeto de un niño que ...             0.59

## Fassttext

No se pudo cargar nunca el modelo, esta muy pesado

https://fasttext.cc/docs/en/crawl-vectors.html

```
!pip install fasttext
```

```
import fasttext.util
import fasttext
import pandas as pd
```

```
model = fasttext.load_model('cc.es.300.bin.gz')

# Calcular la similitud entre frases
def fasttext_similarity(s1, s2):
    # Calcular la similitud coseno entre las representaciones vectoriales de las frases
    similarity = model.cosine_similarity(s1, s2)
    return round(similarity, 2)

# Calcular la similitud para cada par de frases
similarity_scores = [word2vec_similarity(frases_espp['Frase1'].iloc[i], frases_espp['Frase2'].iloc[i]) for i in range(len(frases_espp))]

# Crear un DataFrame con las puntuaciones de similitud
fasttext_p0f = pd.DataFrame({'FastText Similarity': similarity_scores})

# Adjuntar las puntuaciones de similitud al DataFrame original
fasttext_p0f = pd.concat([frases_espp[['Frase1', 'Frase2']], fasttext_p0f], axis=1)
```

## Entrenar un Word2Vec

In [ ]:
!pip install gensim

In [ ]:
import pandas as pd
import numpy as np
from gensim.models import Word2Vec
from nltk.tokenize import word_tokenize
import nltk
nltk.download('punkt')

# Combinar Frase1 y Frase2 en un solo conjunto de datos
all_sentences = pd.concat([frases_espp['Frase1'], frases_espp['Frase2']], ignore_index=True)

# Tokenizar todas las frases
tokenized_sentences = [word_tokenize(sentence.lower()) for sentence in all_sentences]

# Entrenar el modelo Word2Vec
model = Word2Vec(sentences=tokenized_sentences, vector_size=100, window=3, min_count=1, workers=4)

# Guardar el modelo para uso futuro
model.save("word2vec_model.model")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


### Similitud coseno

In [ ]:
# Cargar el modelo entrenado
model = Word2Vec.load("word2vec_model.model")

def word2vec_similarity(s1, s2):
    s1_tokens = word_tokenize(s1.lower())
    s2_tokens = word_tokenize(s2.lower())

    # Filtrar palabras que no están en el vocabulario del modelo
    s1_tokens = [word for word in s1_tokens if word in model.wv.key_to_index]
    s2_tokens = [word for word in s2_tokens if word in model.wv.key_to_index]

    # Calcular la similitud coseno entre los vectores promedio de las palabras
    if s1_tokens and s2_tokens:
        similarity = model.wv.n_similarity(s1_tokens, s2_tokens)
        # Mapear al rango [0, 1]
        scaled_similarity = (similarity + 1) / 2
        return round(scaled_similarity, 2)
    else:
        return 0.0  # Manejar el caso donde ninguna palabra está en el vocabulario

# Calcular la similitud para cada par de frases
similarity_scores = [word2vec_similarity(frases_espp['Frase1'].iloc[i], frases_espp['Frase2'].iloc[i]) for i in range(len(frases_espp))]

# Crear un DataFrame con las puntuaciones de similitud
word2vec_p0c = pd.DataFrame({'Word2Vec Similarity': similarity_scores})

# Adjuntar las puntuaciones de similitud al DataFrame
word2vec_p0c = pd.concat([frases_espp[['Frase1', 'Frase2']], word2vec_p0c], axis=1)

In [ ]:
word2vec_p0c.head(20)

Frase1  \
0              notable es la carta de robert a sarah.   
1   este tipo de tratamiento para grandes contribu...   
2   el filme es denso y con lentitud nos sumerge e...   
3   un " catéter de diálisis " es un catéter usado...   
4   aunque tengamos las mejores intenciones, puede...   
5   shacl (shapes constraint language) es una espe...   
6                 ¿dónde está natchitoches, luisiana?   
7       actualmente forma parte de la " ruta moche ".   
8   no es un robot con personalidad propia, pero s...   
9   mansour bahrami y henri leconte ganaron en la ...   
10  las fibras se utilizan a menudo en la fabricac...   
11  más de 2.000 barcos cargados de esclavos hicie...   
12  ha pasado un mes del terremoto que mató a más ...   
13                 lo saben prácticamente todo de ti.   
14                     si ahora puede demostrarlo ...   
15  que la lluvia iba a llegar era algo que todo e...   
16  en 2017, fue nominada para un sobey art award ...   
17  3 jugadores de baloncesto están tratando de ga...   
18  " quemú quemú " es una localidad ubicada al no...   
19   ¿hay alguna razón por la que ella no te lo dijo?   

                                               Frase2  Word2Vec Similarity  
0   strathairn asistió a williams college, william...                 0.94  
1   bill clinton fue el primer presidente negro de...                 0.93  
2   el talento de johnny depp queda maquillado y d...                 0.96  
3   si un paciente requiere terapia de diálisis de...                 0.97  
4   ", le pregunto a krista fransman, una amiga fi...                 0.93  
5   celebra a lo largo del año diversas festividades.                 0.89  
6            ¿dónde está el puente sobre el río kwai?                 0.90  
7   es el balneario más visitado de la ciudad de t...                 0.95  
8   puedes decirlos en alto o escribirlos, lo que ...                 0.94  
9   jeremy bates y anders järryd ganaron contra ma...                 0.99  
10  los materiales de ingeniería más resistentes a...                 0.96  
11  fue el nacimiento del trip-hop y del sonido br...                 0.93  
12  hoy, la asignación de recursos a hospitales pú...                 0.98  
13  allí llegó después de estudiar filosofía en la...                 0.91  
14           así es como, mmm..., me quedo abrochado.                 0.73  
15  algunos pilotos salieron disparados y consigui...                 0.96  
16  fue nominada para un sobey art award (nominada...                 0.97  
17  dos equipos de jugadores de fútbol no comienza...                 0.96  
18  localidad ubicada al oeste de la llamada pampa...                 0.95  
19  ¿cómo gana un padre el respeto de un niño que ...                 0.91

# Herramientas :

## TF-IDF

Como funciona TfidfVectorizer

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Ejemplo de documentos de texto
documents = ["Este es el primer documento.",
             "Este documento es el segundo documento.",
             "Y este es el tercer documento.",
             "¿Es este el primer documento?"]

# Crear un objeto TfidfVectorizer con tokenización y bigramas
vectorizer = TfidfVectorizer(ngram_range=(1, 2))

# Ajustar y transformar los documentos
tfidf_matrix = vectorizer.fit_transform(documents)

# Obtener el vocabulario (términos) y la matriz TF-IDF
feature_names = vectorizer.get_feature_names_out()
print("Vocabulario (términos):", feature_names)
print("Matriz TF-IDF:\n", tfidf_matrix.toarray())

Vocabulario (términos): ['documento' 'documento es' 'el' 'el primer' 'el segundo' 'el tercer' 'es'
 'es el' 'es este' 'este' 'este documento' 'este el' 'este es' 'primer'
 'primer documento' 'segundo' 'segundo documento' 'tercer'
 'tercer documento']
Matriz TF-IDF:
 [[0.26147516 0.         0.26147516 0.39504337 0.         0.
  0.26147516 0.31982166 0.         0.26147516 0.         0.
  0.39504337 0.39504337 0.39504337 0.         0.         0.
  0.        ]
 [0.38592401 0.36977131 0.19296201 0.         0.36977131 0.
  0.19296201 0.23602023 0.         0.19296201 0.36977131 0.
  0.         0.         0.         0.36977131 0.36977131 0.
  0.        ]
 [0.23066234 0.         0.23066234 0.         0.         0.44201611
  0.23066234 0.28213316 0.         0.23066234 0.         0.
  0.34849058 0.         0.         0.         0.         0.44201611
  0.44201611]
 [0.23445436 0.         0.23445436 0.35421965 0.         0.
  0.23445436 0.         0.44928272 0.23445436 0.         0.44928272
  0.   

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Ejemplo de documentos de texto que incluyen signos de puntuación
documents = ["Este es un ejemplo de texto con signos de puntuación.",
             "¡Los signos de puntuación también son importantes!"]

# Usar TfidfVectorizer con token_pattern personalizado que incluye signos de puntuación
vectorizer = TfidfVectorizer(token_pattern=r'\b\w+\b|[\.,;\'"¡!\?]',  # Incluye signos de puntuación como tokens
                             stop_words='english')  # Opcional: quitar stop words en inglés

# Ajustar y transformar los documentos
tfidf_matrix = vectorizer.fit_transform(documents)

# Obtener el vocabulario (términos) y la matriz TF-IDF
feature_names = vectorizer.get_feature_names_out()
print("Vocabulario (términos):", feature_names)
print("Matriz TF-IDF:\n", tfidf_matrix.toarray())

Vocabulario (términos): ['!' '.' 'ejemplo' 'es' 'este' 'importantes' 'los' 'puntuación' 'signos'
 'son' 'también' 'texto' '¡']
Matriz TF-IDF:
 [[0.         0.4078241  0.4078241  0.4078241  0.4078241  0.
  0.         0.29017021 0.29017021 0.         0.         0.4078241
  0.        ]
 [0.37762778 0.         0.         0.         0.         0.37762778
  0.37762778 0.26868528 0.26868528 0.37762778 0.37762778 0.
  0.37762778]]


## Stemming con NLTK

### Ingles

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize

# Ejemplo de documentos de texto
documents = ["running runner ran", "cook, cooking, cooker, cookies, cake"]

# Stemming con NLTK
stemmer = PorterStemmer()
stemmed_documents = [' '.join([stemmer.stem(word) for word in word_tokenize(doc)]) for doc in documents]

# Usar TfidfVectorizer después del stemming
vectorizer = TfidfVectorizer(token_pattern=r'\b\w+\b|[\.,;\'"¡!\?]')
tfidf_matrix = vectorizer.fit_transform(stemmed_documents)

# Obtener el vocabulario (términos) y la matriz TF-IDF
feature_names = vectorizer.get_feature_names_out()
print("Vocabulario (términos):", feature_names)
print("Matriz TF-IDF:\n", tfidf_matrix.toarray())

Vocabulario (términos): [',' 'cake' 'cook' 'cooker' 'cooki' 'ran' 'run' 'runner']
Matriz TF-IDF:
 [[0.         0.         0.         0.         0.         0.57735027
  0.57735027 0.57735027]
 [0.83405766 0.20851441 0.41702883 0.20851441 0.20851441 0.
  0.         0.        ]]


### Español

In [ ]:
!pip install nltk

In [ ]:
import nltk
from nltk.stem import SnowballStemmer
from nltk.tokenize import word_tokenize
nltk.download('punkt')

# Configurar el stemmer para español
stemmer = SnowballStemmer('spanish')

# Ejemplo de documentos de texto en español
documents = ["corriendo corredor corrió",
             "jugando jugó jugará"]

# Aplicar stemming a cada documento
stemmed_documents = [' '.join([stemmer.stem(word) for word in word_tokenize(doc)]) for doc in documents]

print(stemmed_documents)

['corr corredor corr', 'jug jug jug']


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


#### Con puntuación

In [ ]:
import pandas as pd
import nltk
from nltk.stem import SnowballStemmer
from nltk.tokenize import RegexpTokenizer

# Descargar el recurso necesario para NLTK
nltk.download('punkt')

# Crear un DataFrame de ejemplo
data = {'Texto': ["corriendo, corredor corrió!", "jugando jugó jugará."]}
df = pd.DataFrame(data)

# Configurar el stemmer para español
stemmer = SnowballStemmer('spanish')

# Tokenización con una expresión regular que incluye signos de puntuación
tokenizer = RegexpTokenizer(r'\w+|[^\w\s]')
df['Texto_stemmed'] = df['Texto'].apply(lambda x: ' '.join([stemmer.stem(word) for word in tokenizer.tokenize(x)]))

# Mostrar el DataFrame resultante
print(df)

                         Texto           Texto_stemmed
0  corriendo, corredor corrió!  corr , corredor corr !
1         jugando jugó jugará.           jug jug jug .


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


## Lematización con spaCy

### Inglés

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
import spacy

# Ejemplo de documentos de texto
documents = ["running runner ran", "cook, cooking, cooker, cookies, cake"]

# Lematización con spaCy
nlp = spacy.load("en_core_web_sm")
lemmatized_documents = [' '.join([token.lemma_ for token in nlp(doc)]) for doc in documents]

# Usar TfidfVectorizer después de la lematización
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(lemmatized_documents)

# Obtener el vocabulario (términos) y la matriz TF-IDF
feature_names = vectorizer.get_feature_names_out()
print("Vocabulario (términos):", feature_names)
print("Matriz TF-IDF:\n", tfidf_matrix.toarray())


Vocabulario (términos): ['cake' 'cook' 'cooker' 'cookie' 'cooking' 'run' 'runner']
Matriz TF-IDF:
 [[0.         0.         0.         0.         0.         0.89442719
  0.4472136 ]
 [0.4472136  0.4472136  0.4472136  0.4472136  0.4472136  0.
  0.        ]]


### Español

In [ ]:
!pip install spacy
!python -m spacy download es_core_news_sm

2023-11-28 05:16:40.027666: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-11-28 05:16:40.030147: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-11-28 05:16:40.030473: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-11-28 05:16:44.075712: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.9/12.9 MB 18.9 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_sm')


In [ ]:
import spacy

# Cargar el modelo de spaCy para español
nlp = spacy.load("es_core_news_sm")

# Ejemplo de documentos de texto en español
documents = ["corriendo corredor corrió",
             "jugando jugó jugará"]

# Aplicar lematización a cada documento
lemmatized_documents = [' '.join([token.lemma_ for token in nlp(doc)]) for doc in documents]

print(lemmatized_documents)

['correr corredor correr', 'jugar jugar jugar']


#### Con puntuación

In [ ]:
import pandas as pd
import spacy

# Cargar el modelo de spaCy para español
nlp = spacy.load("es_core_news_sm")

# Crear un DataFrame de ejemplo
data = {'Frase': ["Este es un ejemplo de lematización.",
                  "¡Los signos de puntuación también son importantes!"]}
df = pd.DataFrame(data)

# Aplicar lematización a la columna 'Frase' incluyendo signos de puntuación
df['Frase_lemmatized'] = df['Frase'].apply(lambda x: ' '.join([token.lemma_ for token in nlp(x)]))

# Mostrar el DataFrame resultante
print(df)

                                               Frase  \
0                Este es un ejemplo de lematización.   
1  ¡Los signos de puntuación también son importan...   

                                    Frase_lemmatized  
0             este ser uno ejemplo de lematización .  
1  ¡ el signo de puntuación también ser importante !  
